core/src/main/java/org/elasticsearch/index/seqno/LocalCheckpointTracker.java

```diff
@@ -95,29 +97,30 @@ public LocalCheckpointTracker(final IndexSettings indexSettings, final long maxS
      *
      * @return the next assigned sequence number
      */
-    synchronized long generateSeqNo() {
-        return nextSeqNo++;
+    long generateSeqNo() {
+        return nextSeqNo.getAndIncrement();
     }
 
     /**
      * Marks the processing of the provided sequence number as completed as updates the checkpoint if possible.
      *
      * @param seqNo the sequence number to mark as completed
      */
-    public synchronized void markSeqNoAsCompleted(final long seqNo) {
+    public void markSeqNoAsCompleted(final long seqNo) {
         // make sure we track highest seen sequence number
-        if (seqNo >= nextSeqNo) {
-            nextSeqNo = seqNo + 1;
-        }
+        nextSeqNo.updateAndGet((current) -> seqNo >= current ? seqNo + 1 : current);
         if (seqNo <= checkpoint) {
```

Comment:
I think declaring the whole method as `synchronized` does not help us here because other accesses to `nextSeqNo` are not synchronized anymore and we would then have a concurrency issue (unsafe-check-then-act), no?